### ISSUE TITLE
[RFC] New tags for "boxing" text

### ISSUE THREAD

**Issue by wm4** [libass/libass#375]
I'm suggesting the following tag:
```
\box(x0, y0, x1, y1)
```
This shall be somewhat similar to `\pos(x0, y0)`, but the text is line-broken to not exceed the box width as set by `x1 - x0`. If adding another line would exceed the box height, the text is ellipsized (`Foobar` becomes `Fo...` to fit the width).

It might also be possible to correctly align text to the right for RTL text.

I want to use this for GUI purposes because mpv unfortunately uses libass as rendering API for GUIs. That would be nicer than clipping or guessing.

Thoughts? Not sure if I can/will implement this.

**Comment by astiob** [comment]
It should be currently possible to achieve all of this except the ellipsization by creating a style with the appropriate margins.

I’m wary of ellipsization because it seems like a pretty language-specific thing. Of course, it’s also a thing that’s probably done a million times over in all kinds of software without any regard for language. Users may be used to it. They may not think well of it nevertheless. Heck, even in European languages I often don’t think well of it, although I’ve kinda gotten used to it over the years.

If you really want to avoid the text from overflowing the box, perhaps an alternative to ellipsization could be scrolling the text up and down repeatedly. In fact, doesn’t ASS already have a “banner” feature for this? But it can probably only scroll one way, not there and back again repeatedly.

**Comment by wm4** [comment]
Margins have at least one problem: top and bottom margins are forced to the same value.

Instead of ellipsization you could do a "fade out" (static, not animated). Firefox and Chrome do this for tabs now, I think. Maybe it would be simpler to implement too.

Animating it with scrolling would be weird, I think.

**Comment by astiob** [comment]
> Margins have at least one problem: top and bottom margins are forced to the same value.

Yes, but _if_ you forget about ellipsization/overflow, then IIRC this doesn’t actually matter. Depending on the alignment your text will be positioned at the top or bottom margin and flow freely towards the other end of the screen without regard for the opposite margin, right? Or if it’s explicitly positioned, the vertical margins won’t matter at all.

Also, we could implement support for [ASSv2 syntax, which allows separate top and bottom margins](https://github.com/Cyberbeing/xy-VSFilter/blob/3.1.0.752/src/subtitles/STS.cpp#L1462).

**Comment by wm4** [comment]
Well, I want to cut off the text. One could use \clip, but that won't look very nice. The whole purpose is to constrain text to a specific rectangle, even if the text is not statically known (such as user input or so).

I'm surprised there's an "ASSv2" which vsfilter supports, but we do not.

**Comment by rcombs** [comment]
Hah, I knew v4.00++ meant _something_. Seems pretty easy to add?
Meanwhile, I've long supported the concept of making all ASS header features available as tags in a future format version, and I don't think there's much reason not to make them available in current-versioned scripts when they're just new ways to access existing functionality, so `\margin(left, right, top, bottom)` (with possible simplified forms like `\margin(left, right, V)` and `\margin(H, V)`) seems fine by me.

**Comment by wm4** [comment]
Hm I support such a margin tag. Though I'm not sure whether the simplified forms are worth it, since the ASS event already carries the L/R/V margin values.

**Comment by rcombs** [comment]
I just mean e.g. `\margin(H, V)` behaving the same way as `\margin(H, H, V, V)`. We've got a few other tags that behave similarly.
